Reference content outline (#128)
dojo/site: content/reference-guides/outline/supplemental.md

### Theming

One way applications provide optimal user experiences is via a consistent presentation to end users. This may be as simple as using a consistent font family across similar elements, but often extends to presenting the application in a corporate color palette, or even implementing an entire design language such as Material Design.

Dojo’s styling pipeline makes use of CSS modules to encapsulate style rules to specific widgets and avoid cross-contamination across large codebases. Styles are not entirely isolated however - centralized CSS variables can define common theme attributes and be shared between all application widgets. Custom theming can also be provided for Dojo’s widget suite.

See the [Styling and Theming] reference section for information on how to create application themes.

### User Interface Widgets

Dojo provides several off-the-shelf user interface components through its [widget suite]. Developers can make immediate use widgets that address many common user interface patterns such as comboboxes, buttons, lists, tabs, text input and calendar entry widgets amongst others.

> Dojo's widgets natively support [internationalization, accessibility](#accessibility-a11y-and-internationalization-i18n) and [theming](#theming), giving developers flexibility in delivering user experiences that are unique to their application without needing to invest in custom element creation.

Author: Custom widgets - I've rephrased this to hopefully clarify a bit better

#### Routing in Dojo

Dojo's routing system allows applications to register URL subpaths as routes that link to a specific type of widget, called an Outlet. When a user navigates to a particular route, the outlet widget registered against the route will be rendered.

While outlets are ‘rendered’ when users navigate to them, they seldom deal directly with the rendering of application functionality. Outlets are primarily wrappers that handle navigational concerns - passing of query parameters, or handling error fallbacks - and instead delegate to other widgets within an application for functional rendering.

Applications can provide navigation options to users via Link widgets which are associated to Outlets, in a similar manner to using anchors in traditional HTML pages.

When using routing, Dojo’s build system can [automatically generate separate bundles](#automated-layering) for each top level route within the application. Each bundle can then be independently delivered to users as they are needed.

> See the [Routing] reference section for details on how to implement routing within your application.

Reviewer: ```suggestion
See the [Dojo Routing reference guide] for details on how to implement routing within your application.
```

Dojo allows for easy use of message bundles to separate text messages from application logic, and can optionally make use of relevant portions of [Unicode CLDR] data to support more advanced value formatting where required.

Developing for the web requires applications that are inclusive of users regardless of their accessibility needs. W3C's [accessibility initiative](https://www.w3.org/WAI/) has helped standardize many such requirements, including extra consideration for [Accessible Rich Internet Applications](https://www.w3.org/WAI/standards-guidelines/aria/).

Developing applications using Dojo's [widget suite] provides [WAI-ARIA] attributes out-of-the-box. While Dojo helps in this regard, it can only do so much - application authors have extra responsibility to validate the level of accessibility that their application provides. It is recommended that an explicit accessibility testing step be included in the application's delivery lifecycle.

> See the [Internationalization] reference for more information on how to develop Dojo applications for a global audience.

Reviewer: ```suggestion
See the [Dojo Internationalization reference guide] for more information on how to develop Dojo applications for a global audience.
```

### Adaptable Presentation

With the importance of the internet in modern society, web applications have required to adapt to a proliferation of ways that users access the web. Smaller form-factor mobile experiences have eclipsed desktop usage but larger presentation formats remain valid to deliver complex application requirements. Dojo provides a variety of solutions to help developers create applications that adapt to their users' access needs.

> When pre-rendered content is needed (such as when developing static sites), Dojo applications can make use of build-time rendering ([BTR]) where some or all of the application structure is computed at build time rather than at runtime within a user's browser. Dojo provides a flexible block-based [BTR] solution to support this. that also allows for progressive hydration to support dynamic behavior on top of pre-rendered content. Developers can make use of BTR in this way to optimize the initial set of content and application assets that are delivered to users when they initially access an application.

Reviewer: ```suggestion
When pre-rendered content is needed (such as when developing static sites), Dojo applications can make use of build-time rendering ([BTR]) where some or all of the application structure is computed at build time rather than at runtime within a user's browser. Dojo provides a flexible block-based [BTR] solution to support this. BTR also allows for progressive hydration to support dynamic behavior on top of pre-rendered content. Developers can make use of BTR in this way to optimize the initial set of content and application assets that are delivered to users when they initially access an application.
```

Not certain that this is the right missing word, just that a word was missing.

Author: The whole paragraph was a bit poor - hopefully more informative now.

> Progressive web applications ([PWAs]) can help deliver experiences that are closer to native device apps, while still benefiting from the portability benefits that that web helps support. Dojo helps with the creation of [PWAs] through simple build configuration that enables developers to add features such as offline usage, background data syncing and push notifications to their applications.

Reviewer: "portability benefits that that web helps support" ?

Author: Added further clarity to the whole sentence.

## Application Development Lifecycle

Dojo provides an end-to-end pipeline for developing web applications. Application authors can quickly create new Dojo applications via the [`dojo create app`] CLI command. Applications can then be built in both development and production modes using the [`dojo build app`] command. The build tool allows for rapid development and validation iteration through serving a local HTTP server and watching for further changes to project files. Using this mechanism, developers can make changes and see immediate results in a fully running application.

These commands form part of a modular [Dojo CLI] toolchain that support a variety of uses across the development lifecycle. Applications can configure their build pipeline through a `.dojorc` configuration file in the project root.

> See the [Build] reference section for more information on how to build a variety of applications using Dojo.

Reviewer: ```suggestion
See the [Dojo Build reference guide] for more information on how to build a variety of applications using Dojo.
```

### Testing Strategies

> Not all errors can be caught through compilers or static type checkers. Features can be written that are syntactically and logically valid, but either don’t anticipate problems at runtime, or do not perform functionality in the intended way. To mitigate this risk, additional testing needs to be performed.

Reviewer: ```suggestion
Not all errors can be caught through compilers or static type checkers. Features can get written that are syntactically and logically valid, but either don’t anticipate problems at runtime, or do not perform functionality in the intended way. To mitigate this risk, additional testing needs to be performed.
```

When using the [Dojo CLI] to scaffold applications, a test runner for the [Intern] test library is included by default. This allows developers to begin writing test code immediately alongside application functionality.

Intern provides solutions for many testing concerns but may not be sufficient for all testing needs of a project. Dojo also provides a simple testing harness that allows application test code to validate use of the framework and widgets at the VDOM abstraction level. This harness can be used from any test runner such as Intern, Jest, or any others that an application’s testing strategy necessitates.

> See the [Testing] reference for more details on how to effectively test Dojo applications.

Reviewer: ```suggestion
See the [Dojo Testing reference guide] for more details on how to effectively test Dojo applications.
```

[Dojo CLI]: https://github.com/dojo/cli/blob/master/README.md
[Creating Widgets]: https://github.com/dojo/framework/blob/master/docs/en/creating-widgets/supplemental.md#introduction-to-widgets
[virtualized DOM]: https://github.com/dojo/framework/blob/master/docs/en/creating-widgets/supplemental.md#working-with-the-vdom
[rendering]: https://github.com/dojo/framework/blob/master/docs/en/creating-widgets/supplemental.md#rendering-to-the-dom
[properties]: https://github.com/dojo/framework/blob/master/docs/en/creating-widgets/#node-properties
[styling and theming]: https://github.com/dojo/framework/blob/master/docs/en/styling-and-theming/introduction.md
[basic state management]: https://github.com/dojo/framework/blob/master/docs/en/creating-widgets/supplemental.md#basic-self-encapsulated-widget-state
[properties interface]: https://github.com/dojo/framework/blob/master/docs/en/creating-widgets/supplemental.md#intermediate-passing-widget-properties
[Stores]: https://github.com/dojo/framework/blob/master/docs/en/stores/introduction.md
[Middleware]: https://github.com/dojo/framework/blob/master/docs/en/middleware/introduction.md
[Intern]: https://theintern.io/

> [Routing]: https://github.com/dojo/framework/blob/master/docs/en/routing/introduction.md

Reviewer: ```suggestion
[Dojo Routing reference guide]: https://github.com/dojo/framework/blob/master/docs/en/routing/introduction.md
```

[WAI-ARIA]: https://www.w3.org/TR/wai-aria/

> [Internationalization]: https://github.com/dojo/framework/blob/master/docs/en/i18n/introduction.md

Reviewer: ```suggestion
[Dojo Internationalization reference guide]: https://github.com/dojo/framework/blob/master/docs/en/i18n/introduction.md
```

> [Testing]: https://github.com/dojo/framework/blob/master/docs/en/testing/introduction.md

Reviewer: ```suggestion
[Dojo Testing reference guide]: https://github.com/dojo/framework/blob/master/docs/en/testing/introduction.md
```

> [Build]: https://github.com/dojo/framework/blob/master/docs/en/building/introduction.md

Reviewer: ```suggestion
[Dojo Build reference guide]: https://github.com/dojo/framework/blob/master/docs/en/building/introduction.md
```
